### ISSUE TITLE
Overflow in Mixed Precision training

### ISSUE THREAD

**Issue by MaveriQ** [NVIDIA/apex#238]
I am trying to use Apex Mixed Precision training with a mix of convolution and batchnorm layers; but I get overflow error in optimizer.step() till the dynamic scaling goes down to 1.0 and then it stays at 1.0

“OVERFLOW! Skipping step. Attempted loss scale: 1.0, reducing to 1”

I am using the following network

> ResnetDownBlock(
>   (bn1): BatchNorm2d(3, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
>   (lrelu1): LeakyReLU(negative_slope=0.1)
>   (conv1): Conv2d(3, 3, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
>   (bn2): BatchNorm2d(3, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
>   (lrelu2): LeakyReLU(negative_slope=0.1)
>   (conv2): Conv2d(3, 3, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
>   (main): Sequential(
>     (0): BatchNorm2d(3, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
>     (1): LeakyReLU(negative_slope=0.1)
>     (2): Conv2d(3, 3, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
>     (3): BatchNorm2d(3, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
>     (4): LeakyReLU(negative_slope=0.1)
>     (5): Conv2d(3, 3, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
>   )
>   (shortcut): Sequential(
>     (0): Conv2d(3, 3, kernel_size=(1, 1), stride=(2, 2), bias=False)
>     (1): BatchNorm2d(3, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
>     (2): LeakyReLU(negative_slope=0.1)
>   )
> )

Then I convert the network to mixed precision taking care of the batchnorm:

> net=apex.fp16_utils.BN_convert_float(ResnetDownBlock(3,1).cuda().half())
> opt=optim.Adam(net.parameters())
> opt=apex.fp16_utils.FP16_Optimizer(opt,dynamic_loss_scale=True)
> 
> pred=net(torch.rand(8,3,2,2).cuda().half()).squeeze()
> loss=nn.MSELoss()(pred.float(),torch.ones((8)).cuda())
> opt.backward(loss)
> opt.step()

The last step produces the warning : “OVERFLOW! Skipping step. Attempted loss scale: 4294967296, reducing to 2147483648.0” which reduces to 1.0

I looked up the network parameters if they are converted to mixed precision appropriately and I get following :

for tag,param in net.named_parameters():
print(‘Parameter: {}\t Datatype: {}’.format(tag,param.dtype))

> |Parameter: bn1.weight| Datatype: torch.float32|
> |Parameter: bn1.bias| Datatype: torch.float32|
> |Parameter: conv1.weight| Datatype: torch.float16|
> |Parameter: bn2.weight| Datatype: torch.float32|
> |Parameter: bn2.bias| Datatype: torch.float32|
> |Parameter: conv2.weight| Datatype: torch.float16|
> |Parameter: shortcut.0.weight| Datatype: torch.float16|
> |Parameter: shortcut.1.weight| Datatype: torch.float32|
> |Parameter: shortcut.1.bias| Datatype: torch.float32|

Apparently this is correct, but I can’t get rid of the overflow.

I posted it on PyTorch forums 2 months back but didn't get any help there (https://discuss.pytorch.org/t/overflow-in-mixed-precision-training-using-apex/35229)

Thanks

**Comment by mcarilli** [comment]
You're using the old (deprecated) FP16_Optimizer and BN_convert_float APIs, which we no longer support.  Documentation for the new API can be found here:  https://nvidia.github.io/apex/amp.html

Try reinstalling a fresh clone of master
```
$ pip uninstall apex
$ pip unistall apex (repeat til you're sure it's gone)
$ cd apex_repo_dir
$ git pull
$ pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" .
```
then modify your code as follows:
```
    # build net in default FP32, don't worry about anything mixed-precision-related
    net = net.cuda() # amp.initialize expects that your model is already cuda()ed as desired
    opt=optim.Adam(net.parameters())

    net, opt = amp.initialize(net, opt, opt_level="O1")
    # If you wish to wrap your model with torch.nn.parallel.DistributedDataParallel or
    # apex.parallel.DistributedDataParallel, do it here, AFTER amp.initialize
 
    # Also note that we don't need to call .half() on the input manually here.
    # Amp patches the model's forward() method to cast input for you if needed.
    pred=net(torch.rand(8,3,2,2).cuda()).squeeze()
    loss=nn.MSELoss()(pred,torch.ones((8)).cuda())

    with amp.scale_loss(loss, opt) as scaled_loss:
        scaled_loss.backward()

    opt.step()
```
`opt_level="O1"` (which implements mixed precision by patching the torch namespace) is recommended to begin with, but `opt_level="O2"` is roughly equivalent to the old thing called FP16_Optimizer, so you can try that as well (nothing else needs to change other than the `opt_level` flag itself).  A detailed explanation of opt_levels can be found in the docs link above.

Also, running your network over and over again on new random input data each iteration might be fundamentally unstable.  What is it even trying to learn in such a case?

**Comment by MaveriQ** [comment]
Thank you for quick reply. I will install Apex again and get back to you. 

I used torch.rand(8,3,2,2) as place-holder for actual input for the network; only for the purpose of this demo. I was getting the same overflow on actual images.

**Comment by MaveriQ** [comment]
Thank you. Works like a charm! And the new API and documentation is much cleaner and intuitive!
